File: Source/GITObjectHash.m
//
//  GITObjectHash.m
//  Git.framework
//
//  Created by Geoff Garside on 08/11/2009.
//  Copyright 2009 Geoff Garside. All rights reserved.
//

#import "GITObjectHash.h"


const NSUInteger GITObjectHashLength = 40;
const NSUInteger GITObjectHashPackedLength = 20;
const NSString *hexChars = @"0123456789abcdef";

@implementation GITObjectHash

@synthesize data;

- (id)initWithData:(NSData *)theData {
    self = [super init];
    if ( self ) {
        self.data = theData;
    }
    return self;
}

- (id)initWithString:(NSString *)aString {
    self = [super init];
    if ( self ) {
        self.data = [[self class] packedDataFromString:aString];
    }
    return self;
}

- (id)initWithBytes:(uint8_t *)bytes length:(size_t)length {
    self = [super init];
    if ( self ) {
        NSData *bytesData = [NSData dataWithBytes:bytes length:length];
        if ( length == GITObjectHashLength ) {
            NSString *string = [[NSString alloc] initWithData:bytesData encoding:NSASCIIStringEncoding];
            self.data = [[self class] packedDataFromString:string];
            [string release];
        }
        else if ( length == GITObjectHashPackedLength ) {
            self.data = bytesData;
        }
    }
    return self;
}

+ (GITObjectHash *)objectHashWithString:(NSString *)aString {
    return [[[self alloc] initWithString:aString] autorelease];
}

+ (GITObjectHash *)objectHashWithData:(NSData *)theData {
    return [[[self alloc] initWithData:theData] autorelease];
}

//! \name Packing and Unpacking SHA1 Hashes
+ (NSData *)packedDataFromString:(NSString *)aString {
    NSUInteger length = [aString length];
    if ( length != GITObjectHashLength )
        return nil;
    
    NSUInteger i;
    NSRange range;
    uint8_t raw[GITObjectHashPackedLength];
    
    // In a hex string, each character represents 4 bits.
    // We go through the string 2 characters at a time to form one byte per loop.
    for ( i = 0; i < length; i += 2 ) {
        uint8_t *current = raw + i/2;
        range = [hexChars rangeOfString:[aString substringWithRange:NSMakeRange(i, 1)]];
        *current = range.location << 4;
        range = [hexChars rangeOfString:[aString substringWithRange:NSMakeRange(i + 1, 1)]];
        *current |= range.location;
    }
    return [NSData dataWithBytes:raw length:GITObjectHashPackedLength];
}

+ (NSString *)unpackedStringFromData:(NSData *)theData {
    NSUInteger length = [theData length];
    if ( length != GITObjectHashPackedLength )
        return nil;
    
    NSUInteger i;
    NSMutableString *string = [NSMutableString string];
    const uint8_t *raw = [theData bytes];
    
    // In a hex string, each character represents 4 bits.
    // We go through all bytes and add 2 hex characters for each byte.
    for ( i = 0; i < length; i++ ) {
        [string appendString:[hexChars substringWithRange:NSMakeRange(*raw >> 4, 1)]];
        [string appendString:[hexChars substringWithRange:NSMakeRange(*raw & 0x0f, 1)]];
        raw++;
    }
    return string;
}


//! \name Getting Packed and Unpacked Forms
- (NSString *)unpackedString {
    return [[self class] unpackedStringFromData:data];
}

//! \name Testing Equality
- (BOOL)isEqual:(id)other {
    if ( !other ) return NO;                            // other is nil
    if ( other == self ) return YES;                    // pointers match?

    if ( [other isKindOfClass:[self class]] )           // Same class?
        return [self isEqualToObjectHash:other];
    if ( [other isKindOfClass:[NSString class]] )       // A string?
        return [self isEqualToString:other];
    if ( [other isKindOfClass:[NSData class]] )         // A data?
        return [self isEqualToData:other];

    return NO;                                          // Definitely not then
}

- (BOOL)isEqualToObjectHash: (GITObjectHash *)otherObjectHash {
    if ( !otherObjectHash )
        return NO;
    if ( self == otherObjectHash )
        return YES;
    if ( [data isEqualToData:otherObjectHash.data] )
        return YES;
    return NO;
}

- (BOOL)isEqualToData: (NSData *)theData {
    return [data isEqualToData:theData];
}

- (BOOL)isEqualToString: (NSString *)aString {
    return [[self unpackedString] isEqualToString:aString];
}

- (NSString *)description {
    return [self unpackedString];
}

- (void)dealloc {
    self.data = nil;
    [super dealloc];
}

@end

File: Test/Unit/GITObjectHash.rb
describe "GITObjectHash" do
  before do
    @sha1_str = "bed4001738fa8dad666d669867afaf9f2c2b8c6a"
    @pack_data = [@sha1_str].pack("H*").data
    @hash = GITObjectHash.alloc.initWithString(@sha1_str)
  end

  describe "+packedDataFromString:" do
    before do
      @subj = GITObjectHash.packedDataFromString(@sha1_str)
    end

    should "return NSData" do
      @subj.should.be.kind_of NSData
    end

    should "return data 20 bytes long" do
      @subj.length.should == GITObjectHashPackedLength
    end

    should "return data matching sha1_data" do
      @subj.should === @pack_data
    end
  end

  describe "+unpackedStringFromData:" do
    before do
      @subj = GITObjectHash.unpackedStringFromData(@pack_data)
    end

    should "return NSString" do
      @subj.should.be.kind_of NSString
    end

    should "return a string 40 characters long" do
      @subj.length.should == GITObjectHashLength
    end

    should "return data matching pack_data" do
      @subj.should === @sha1_str
    end
  end

  describe "-unpackedString" do
    should "return NSString" do
      @hash.unpackedString.should.be.kind_of NSString
    end

    should "return a 40 character long string" do
      @hash.unpackedString.length.should == GITObjectHashLength
    end

    should "return the correct string" do
      @hash.unpackedString.should == @sha1_str
    end
  end

  describe "-packedData" do
    should "return NSData" do
      @hash.packedData.should.be.kind_of NSData
    end

    should "return 20 bytes of data" do
      @hash.packedData.length.should == GITObjectHashPackedLength
    end

    should "return the correct data" do
      @hash.packedData.should === @pack_data
    end
  end

  describe "-isEqual:" do
    before do
      @other = GITObjectHash.objectHashWithData(@pack_data)
    end

    should "be true with @hash" do
      @other.isEqual(@hash).should.be.true
    end

    should "be true with @sha1_str" do
      @hash.isEqual(@sha1_str).should.be.true
    end

    should "be true with @pack_data" do
      @hash.isEqual(@pack_data).should.be.true
    end

   should "be false with 1" do
     @hash.isEqual(1).should.be.false
   end

   should "be false with 'hello'" do
     @hash.isEqual("hello").should.be.false
   end
  end

  describe "-isEqualToObjectHash:" do
    should "be true with @hash" do
      GITObjectHash.objectHashWithData(@pack_data).isEqualToObjectHash(@hash).should.be.true
    end
  end

  describe "-isEqualToData:" do
    should "be true with @pack_data" do
      @hash.isEqualToData(@pack_data).should.be.true
    end
  end

  describe "-isEqualToString:" do
    should "be true with @sha1_str" do
      @hash.isEqualToString(@sha1_str).should.be.true
    end
  end
end
